File: dms-ai-engine/services/escalation_tracker.py
class EscalationTracker:
    """
    Tính escalation_level (1/2/3) dựa trên thời gian LIÊN TỤC ở CRITICAL đã qua
    hysteresis (driven bởi critical_active property của TriggerEmitter/
    DistractionTriggerEmitter -- KHÔNG phải raw score > threshold từng frame).
    """

    def __init__(self, level_up_seconds: list[float], repeat_interval_seconds: list[float]):
        assert len(repeat_interval_seconds) == len(level_up_seconds) + 1
        self._level_up_seconds = level_up_seconds
        self._repeat_interval_seconds = repeat_interval_seconds
        self._critical_since: float | None = None
        self._last_repeat_time: float | None = None
        self._last_level = 1

    def update(self, critical_active: bool, now: float) -> tuple[int, bool, bool]:
        """Trả về (level, repeat_due, level_changed)."""
        if not critical_active:
            self._critical_since = None
            self._last_repeat_time = None
            level_changed = self._last_level != 1
            self._last_level = 1
            return 1, False, level_changed

        if self._critical_since is None:
            self._critical_since = now
            self._last_repeat_time = None  # chưa lặp lần nào trong episode này

        elapsed = now - self._critical_since
        level = 1 + sum(1 for t in self._level_up_seconds if elapsed >= t)

        level_changed = level != self._last_level
        self._last_level = level

        interval = self._repeat_interval_seconds[level - 1]
        # level_changed cũng tự tính là "vừa lặp" -- nếu không, _last_repeat_time
        # cũ (neo theo interval của level TRƯỚC) có thể khiến lần repeat_due kế
        # tiếp rơi chỉ 1-2s sau khi vừa nói câu ở level MỚI, cắt ngang utterance
        # đó giữa chừng.
        repeat_due = level_changed or self._last_repeat_time is None or (now - self._last_repeat_time) >= interval
        if repeat_due:
            self._last_repeat_time = now
        return level, repeat_due, level_changed

    def reset(self) -> None:
        """Gọi khi UNKNOWN (mất mặt) bắn ra -- force về level 1 bất kể
        critical_active hiện tại (TriggerEmitter không cập nhật trong lúc mất
        mặt, nên không thể tự phản ánh qua critical_active)."""
        self._critical_since = None
        self._last_repeat_time = None
        self._last_level = 1

File: aaos-cockpit-app/app/src/main/java/com/vitalguard/ai/drowsiness/EscalationTracker.kt
/**
 * Ported from dms-ai-engine/services/escalation_tracker.py. Computes an
 * escalation level (1/2/3) from how long the underlying emitter's
 * criticalActive has been continuously true -- driven by TriggerEmitter's
 * hysteresis-protected criticalActive, NOT a raw score>threshold check per
 * frame.
 */
package com.vitalguard.ai.drowsiness

/**
 * Ported from dms-ai-engine/services/escalation_tracker.py. Computes an
 * escalation level (1/2/3) from how long the underlying emitter's
 * criticalActive has been continuously true -- driven by TriggerEmitter's
 * hysteresis-protected criticalActive, NOT a raw score>threshold check per
 * frame.
 */
class EscalationTracker(
    private val levelUpSeconds: List<Double>,
    private val repeatIntervalSeconds: List<Double>,
) {
    init {
        require(repeatIntervalSeconds.size == levelUpSeconds.size + 1) {
            "repeatIntervalSeconds must have exactly one more entry than levelUpSeconds"
        }
    }

    private var criticalSince: Double? = null
    private var lastRepeatTime: Double? = null
    private var lastLevel: Int = 1

    /** Returns (level, repeatDue, levelChanged). */
    fun update(criticalActive: Boolean, now: Double): Triple<Int, Boolean, Boolean> {
        if (!criticalActive) {
            criticalSince = null
            lastRepeatTime = null
            val levelChanged = lastLevel != 1
            lastLevel = 1
            return Triple(1, false, levelChanged)
        }

        if (criticalSince == null) {
            criticalSince = now
            lastRepeatTime = null // no repeat yet in this episode
        }

        val elapsed = now - criticalSince!!
        val level = 1 + levelUpSeconds.count { elapsed >= it }

        val levelChanged = level != lastLevel
        lastLevel = level

        val interval = repeatIntervalSeconds[level - 1]
        // levelChanged also counts as "just repeated" -- otherwise a stale
        // lastRepeatTime (anchored to the PREVIOUS level's interval) could
        // make the next repeatDue fire only 1-2s after the level-change
        // announcement, cutting off the utterance just spoken for the new level.
        val repeatDue = levelChanged || lastRepeatTime == null || (now - lastRepeatTime!!) >= interval
        if (repeatDue) lastRepeatTime = now

        return Triple(level, repeatDue, levelChanged)
    }

    /** Called on UNKNOWN (face lost) -- forces back to level 1 regardless of
     * criticalActive, since the underlying emitter is frozen (not updated)
     * while hasFace=false and cannot itself signal a reset. */
    fun reset() {
        criticalSince = null
        lastRepeatTime = null
        lastLevel = 1
    }
}
